### PR TITLE
Add the article about kmcuda v4

### PR DESCRIPTION
**Pull request by vmarkovtsev**

### CODE REVIEW

File: content/post/kmcuda4.md [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+
+I will go into details about each of the points.
+
+### Multi-GPU
+
+![nvprof](/post/kmcuda4/nvprof.png)
+
+Clustering on several GPUs at the same time is not straightforward,
+because you cannot simply split the samples into equal parts and
+feed them in parallel. The calculation of the centroifs requires
```
Comment (alcortesm) [inline content/post/kmcuda4.md]: s/centroifs/centroids/

File: content/post/kmcuda4.md [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+all the data in the same place. While there can be workarounds for
```
Comment (alcortesm) [inline content/post/kmcuda4.md]: add "having" at the begining of this line:

"The calculation of the centroids requires having all the data in the same place."

File: content/post/kmcuda4.md [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+
+1. We calculate the distances between the samples and the centroids,
+pick the centroid with the minimum distance fo every sample (assignment stage).
+2. We update the centroids by averaging all the samples assigned to each
+(centroids stage).
+
+The current multi-gpu scheme is as follows.
+
+1. We cut the samples into the same number of intervals as the number of
+GPUs. Each GPU works with one interval, calculates the distances,
```
Comment (alcortesm) [inline content/post/kmcuda4.md]: consider rewording the first sentence in the paragraph to:

"We cut the samples into as many intervals as GPUs are available."

File: content/post/kmcuda4.md [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+writes the local assignments.
```
Comment (alcortesm) [inline content/post/kmcuda4.md]: change the last "," in the previous line to an "and".

File: content/post/kmcuda4.md [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+        } while (false);
+        (device_ccounts)
+            .emplace_back(
+                reinterpret_cast<std::remove_reference<decltype(
+                    device_ccounts)>::type::value_type::element_type*>(__ptr));
+      } while (false);
+    }
+  } while (false);
+} while (false)
+```
```
Comment (alcortesm) [inline content/post/kmcuda4.md]: holy fuck!, I would not like to debug that macro!

File: content/post/kmcuda4.md [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+
+And this is only one tiny example.
+I am really sorry what I have done. However, I don't
```
Comment (alcortesm) [inline content/post/kmcuda4.md]: This sentence doesn't sound as correct English to me, but I don't know how to improve it. Maybe something along the lines of: "I regret having end up using macros like this"...

**Comment by vmarkovtsev** [comment]
@alcortesm thanks! Corrected everything you noted.

**Pull request merged**